jaxlib/cuda_plugin_extension.cc

```diff
@@ -63,14 +65,73 @@ absl::Status RegisterCustomCallTarget(const PJRT_Api* c_api,
   args.struct_size = PJRT_Gpu_Register_Custom_Call_Args_STRUCT_SIZE;
   args.function_name = fn_name_c_str;
   args.function_name_size = fn_name_size;
+
 #if PJRT_API_GPU_EXTENSION_VERSION >= 1
   args.api_version = api_version;
 #endif
-  args.custom_call_function = static_cast<void*>(fn.data());
-  RETURN_STATUS_IF_PJRT_ERROR(
-      reinterpret_cast<const PJRT_Gpu_Custom_Call*>(next)->custom_call(&args),
-      c_api);
+
+  auto as_capsule = [](nb::object obj) -> absl::StatusOr<nb::capsule> {
+    nb::capsule capsule;
+    if (!nb::try_cast<nb::capsule>(obj, capsule)) {
+      return absl::InvalidArgumentError(
+          "Custom call target registration requires handlers as PyCapsules");
+    }
+    return capsule;
+  };
+
+#if PJRT_API_GPU_EXTENSION_VERSION <= 1
+  TF_ASSIGN_OR_RETURN(nb::capsule fn_execute, as_capsule(fn));
+  args.custom_call_function = fn_execute.data();
+  RETURN_STATUS_IF_PJRT_ERROR(register_custom_call(&args), c_api);
   return absl::OkStatus();
+#else
+  args.handler_instantiate = nullptr;
+  args.handler_prepare = nullptr;
+  args.handler_initialize = nullptr;
+  args.handler_execute = nullptr;
+
+  // Register legacy custom call target (untyped void* API).
+  if (api_version == 0) {
+    TF_ASSIGN_OR_RETURN(nb::capsule capsule_execute, as_capsule(fn));
+    args.handler_execute = capsule_execute.data();
+    RETURN_STATUS_IF_PJRT_ERROR(register_custom_call(&args), c_api);
+    return absl::OkStatus();
+  }
+
+  // Register XLA FFI handler (typed API with explicit function signatures).
+  if (api_version == 1) {
+    auto capsule_execute = as_capsule(fn);
+    if (capsule_execute.ok()) {
+      args.handler_execute = capsule_execute->data();
+      RETURN_STATUS_IF_PJRT_ERROR(register_custom_call(&args), c_api);
+      return absl::OkStatus();
+    }
```

Comment:
This could be consolidated with the `api_version == 0` block above, I think? I don't feel strongly, but it might be useful to avoid duplicated logic.

Reply:
I don't feel strongly about it either -- I would only suggest that whatever case distinction we make here we should more or less keep in sync with https://github.com/openxla/xla/blob/b9fcb2422f5e38ed8aecaec6b604d2fc86755c4d/xla/python/xla_compiler.cc#L306-L308 (which is the case now).

Reply:
That's fair! The difference here is that the logic about which registry to use is abstracted away, meaning that the two code blocks here and above are identical. In `xla_compiler.cc`, the API version does actually change the registration code, so we do need to maintain the two versions over there.

Reply:
Exactly. If in the future we share code for all three of these uses (xla_compiler.cc + GPU plugins) we might want to add a `RegisterCustomCallTarget`-like helper that's parametric in `register_custom_call(fn_name, bundle)`.